src/core/ext/filters/client_channel/channel_connectivity.cc

```diff
@@ -152,15 +152,18 @@ static void partly_done(state_watcher* w, bool due_to_completion,
         w->error = error;
       }
       w->phase = CALLING_BACK_AND_FINISHED;
-      grpc_cq_end_op(w->cq, w->tag, w->error, finished_completion, w,
-                     &w->completion_storage);
       break;
     case CALLING_BACK_AND_FINISHED:
       GPR_UNREACHABLE_CODE(return );
       break;
   }
   gpr_mu_unlock(&w->mu);
 
+  if (w->phase == CALLING_BACK_AND_FINISHED) {
```

Comment:
Generally, if a decision is being made under a lock and not taking effect until after a lock, the best way to represent that is to capture the decision and all its state into local variables while still under the lock, and then do the action using those local variables after the lock. Otherwise, it is likely to lead to races. So I would think a more common use pattern inside the lock would be

```
end_op = true;
end_op_cq = w->cq;
end_op_tag = w->tag;
end_op_error = w->error;
end_op_completion_storage = &w->completion_storage;
```
etc

and then at this point something like

``` 
if (end_op) {
  grpc_cq_end_op(end_op_cq, end_op_tag, end_op_error, finished_completion, w, end_op_completion_storage);
}
```

Reply:
Thanks for the feedback, changed it as you suggested.